nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -547,6 +506,173 @@ public void compareGenericTypeParameterNullabilityForCall(
+          }
+          List<Type> lhsTypeArguments = lhsType.getTypeArguments();
+          List<Type> rhsTypeArguments = rhsType.getTypeArguments();
+          // This is impossible, considering the fact that standard Java subtyping succeeds before
+          // running
+          // NullAway
+          if (lhsTypeArguments.size() != rhsTypeArguments.size()) {
+            throw new RuntimeException(
+                "Number of types arguments in " + rhsType + " does not match " + lhsType);
+          }
+          for (int i = 0; i < lhsTypeArguments.size(); i++) {
+            Type lhsTypeArgument = lhsTypeArguments.get(i);
+            Type rhsTypeArgument = rhsTypeArguments.get(i);
+            boolean isLHSNullableAnnotated = false;
+            List<Attribute.TypeCompound> lhsAnnotations = lhsTypeArgument.getAnnotationMirrors();
+            // To ensure that we are checking only jspecify nullable annotations
+            for (Attribute.TypeCompound annotation : lhsAnnotations) {
+              if (annotation.getAnnotationType().toString().equals(NULLABLE_NAME)) {
+                isLHSNullableAnnotated = true;
+                break;
+              }
+            }
+            boolean isRHSNullableAnnotated = false;
+            List<Attribute.TypeCompound> rhsAnnotations = rhsTypeArgument.getAnnotationMirrors();
+            // To ensure that we are checking only jspecify nullable annotations
+            for (Attribute.TypeCompound annotation : rhsAnnotations) {
+              if (annotation.getAnnotationType().toString().equals(NULLABLE_NAME)) {
+                isRHSNullableAnnotated = true;
+                break;
+              }
+            }
+            if (isLHSNullableAnnotated != isRHSNullableAnnotated) {
+              return false;
+            }
+            // nested generics
+            List<Type> genericArgs = lhsTypeArgument.accept(TYPE_ARG_VISITOR, null);
+            if (genericArgs.size() > 0) {
+              if (!lhsTypeArgument.accept(COMPARE_NULLABILITY_VISITOR, rhsTypeArgument)) {
+                return false;
+              }
+            }
+          }
+          return true;
+        }
+
+        @Override
+        public Boolean visitArrayType(Type.ArrayType lhsType, Type rhsType) {
+          Type.ArrayType arrRhsType = (Type.ArrayType) rhsType;
+          return lhsType
+              .getComponentType()
+              .accept(COMPARE_NULLABILITY_VISITOR, arrRhsType.getComponentType());
+        }
+
+        @Override
+        @Nullable
+        public Boolean visitType(Type t, Type type) {
+          return null;
+        }
+      };
+
+  /** Visitor For Handling Different Tree Types */
+  public class PreservedAnnotationTreeVisitor extends SimpleTreeVisitor<Type, Boolean> {
+    @Override
+    public Type visitArrayType(ArrayTypeTree tree, Boolean p) {
+      ParameterizedTypeTree paramTree = (ParameterizedTypeTree) tree.getType();
```

Comment:
Are we sure `tree.getType()` will always be a `ParameterizedTypeTree`?

Reply:
I am not sure, I tried running a test with a regular ArrayType inside the generic like A<String[]> and that ran fine. Maybe its better to just do `Type.ClassType classType =
    (Type.ClassType) tree.getType().accept(new PreservedAnnotationTreeVisitor(), null);
return new Type.ArrayType(classType, classType.tsym);` instead of assuming?

Reply:
Can you try a type like `A<int[]>`?  If we can't find a counterexample we can leave this code as is

Reply:
It gives a ClassCastException in the TYPE_ARG_VISITOR since it cannot cast PrimitiveType to ClassType

Reply:
Ok then we need a fix somewhere.  Please add that as another test case as well

Reply:
It was a simple fix to not cast it into ClassType there and there was no other issue